### ISSUE TITLE
Use C++ classes as base

### ISSUE THREAD

**Issue by langmm** [cropsinsilico/cis_interface#30]
Create C++ classes replicating major methods except serialization (#72) with Python wrappers (Python API preferred or Cython)

### Communicator (Comm) Classes (in order of priority)
- [ ] Base
- [ ] ZMQ using zeromq C++ library directly (previously used czmq which has added complexity to installation)
- [ ] Async class (can use any base class, but messages are queued and sent/received asynchronously using threads)
- [ ] IPC class using sysv queues (lower priority)
- [ ] RMQ class (lowest priority)

### Methods
- [ ] Constructors/destructors that check for address information in environment variables and manage underlying resources for the base class communication (e.g. queues)
- [ ] Send/Receive small string messages
- [ ] Send/Receive large string messages by breaking up the message and sending the pieces to a temporary comm identified by header information sent to the recipient
- [ ] Open comm, opening connections and altering state that can be checked via method
- [ ] Close comm, closing connections and altering state that can be checked via method
- [ ] Methods for checking if messages are available for receipt or if messages have not been received
- [ ] (AFTER #72 COMPLET) Send/Receive non-string messages

### Additional requirements
- [ ] Tests for C++ classes
- [ ] CI for C++ tests
- [ ] Python wrapper classes serve as drop in replacements for the existing Python comm classes and pass interface & connection driver tests (parts of existing comm tests may be modified where they touch non-stable API).
- [ ] C wrapper
- [ ] Fortran wrapper using ISO_C_BINDING
- [ ] (MAYBE) R/Matlab wrapper directly wrapping C++ instead of Python?
